### ISSUE TITLE
请问怎么获得完整代码

### ISSUE THREAD

**Issue by ws1hope** [GeoZcx/A-deeply-supervised-image-fusion-network-for-change-detection-in-remote-sensing-images#27]
作者你好，在公开代码中的DSFIN脚本中，不太懂model_A,model_B,是用来做什么的，请问有什么方式可以获得完整的代码吗？万分感谢

**Comment by xuican** [comment]
完整的代码有吗  晓晨博士

**Comment by xuican** [comment]
is VGG16？

**Comment by ws1hope** [comment]
No,I need training and prediction scripts to understand model_A and model_B referenced in DSIFN

**Comment by songyvying** [comment]
同求完整代码，希望作者可以分享predict.py部分的代码，万分感谢 @GeoZcx

**Comment by GeoZcx** [comment]
> 作者你好，在公开代码中的DSFIN脚本中，不太懂model_A,model_B,是用来做什么的，请问有什么方式可以获得完整的代码吗？万分感谢

你好，感谢对此工作的关注！model_A和model_B是VGG16 backbone，您看到的应该是pytorch版本的代码，此段代码是我依据keras版本代码改写的，因此没有训练部分的code，从keras版本可以看到backbone是固定不需要更新的，因此这里的pytorch版本有些冗余，这里其实使用单个backbone，也就是一个vgg16 backbone就可以了。训练超参数可以参考百度网盘里的IFN _train_code.py 文件，由于时间久远，pytorch版本训练代码就不再更新了，请见谅！

**Comment by GeoZcx** [comment]
> 

你好，感谢对此工作的关注！我将keras版本的预测代码贴在下面：

from keras.models import model_from_json,load_model
import os
import numpy as np
import rasterio
from losses import dice_loss,balanced_loss,bce_dice_loss
from metrics import f1
from keras.preprocessing import image

os.environ["CUDA_VISIBLE_DEVICES"] = '2'


def read_4bands_image(file_path):
    ras_dataset = rasterio.open(file_path)
    ras_band1 = ras_dataset.read(1)
    ras_band2 = ras_dataset.read(2)
    ras_band3 = ras_dataset.read(3)

    band_stack = np.array([ras_band1, ras_band2, ras_band3])
    band_stack = np.transpose(band_stack, (1, 2, 0))
    list_tc_0 = np.array(band_stack)

    return list_tc_0

def predict():
    train_t1_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t1")
    train_t2_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t2")

    model = load_model('best_model.h5', custom_objects={'bce_dice_loss': bce_dice_loss,'f1':f1})


    list_dir = os.listdir(train_t1_path)

    for i in list_dir:
        t1_data = []
        t2_data = []
        train_t1 = read_4bands_image(os.path.join(train_t1_path,i))/255.0
        train_t2 = read_4bands_image(os.path.join(train_t2_path,i.replace("t1","t2")))/255.0
        t1_data.append(train_t1)
        t2_data.append(train_t2)
        t1_data = np.array(t1_data)
        t2_data = np.array(t2_data)
        # train_change = np.concatenate([t1_data,t2_data],axis=3)

        result = model.predict([t1_data,t2_data])[3]
        predict_path = os.path.join("predicted", i)
        result = result[0,:,:,:]

        result = np.transpose(result, (2,0,1))
        result = result[0]
        result[result<0.5] = 0.0
        result[result >= 0.5] = 255.0

        #save image
        bands_combination = np.zeros(shape=(256,256,3))
        bands_combination[:, :, 0] = result
        bands_combination[:, :, 1] = result
        bands_combination[:, :, 2] = result

        predict_result = image.array_to_img(bands_combination)
        image.save_img(os.path.join('predicted', i), predict_result)

predict()

**Comment by GeoZcx** [comment]
> 同求完整代码，希望作者可以分享predict.py部分的代码，万分感谢 @GeoZcx

你好，感谢对此工作的关注！我将keras版本的预测代码贴在下面：

from keras.models import model_from_json,load_model
import os
import numpy as np
import rasterio
from losses import dice_loss,balanced_loss,bce_dice_loss
from metrics import f1
from keras.preprocessing import image

os.environ["CUDA_VISIBLE_DEVICES"] = '2'


def read_4bands_image(file_path):
    ras_dataset = rasterio.open(file_path)
    ras_band1 = ras_dataset.read(1)
    ras_band2 = ras_dataset.read(2)
    ras_band3 = ras_dataset.read(3)

    band_stack = np.array([ras_band1, ras_band2, ras_band3])
    band_stack = np.transpose(band_stack, (1, 2, 0))
    list_tc_0 = np.array(band_stack)

    return list_tc_0

def predict():
    train_t1_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t1")
    train_t2_path = os.path.join("/home/zcx/ChangeDetectionDataset/test","t2")

    model = load_model('best_model.h5', custom_objects={'bce_dice_loss': bce_dice_loss,'f1':f1})


    list_dir = os.listdir(train_t1_path)

    for i in list_dir:
        t1_data = []
        t2_data = []
        train_t1 = read_4bands_image(os.path.join(train_t1_path,i))/255.0
        train_t2 = read_4bands_image(os.path.join(train_t2_path,i.replace("t1","t2")))/255.0
        t1_data.append(train_t1)
        t2_data.append(train_t2)
        t1_data = np.array(t1_data)
        t2_data = np.array(t2_data)
        # train_change = np.concatenate([t1_data,t2_data],axis=3)

        result = model.predict([t1_data,t2_data])[3]
        predict_path = os.path.join("predicted", i)
        result = result[0,:,:,:]

        result = np.transpose(result, (2,0,1))
        result = result[0]
        result[result<0.5] = 0.0
        result[result >= 0.5] = 255.0

        #save image
        bands_combination = np.zeros(shape=(256,256,3))
        bands_combination[:, :, 0] = result
        bands_combination[:, :, 1] = result
        bands_combination[:, :, 2] = result

        predict_result = image.array_to_img(bands_combination)
        image.save_img(os.path.join('predicted', i), predict_result)

predict()
